pkg/kubelet/cm/dra/manager.go

```diff
@@ -108,48 +108,73 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
 			continue
 		}
 
-		claimInfo := m.cache.get(*claimName, pod.Namespace)
-		if claimInfo == nil {
-			// claim does not exist in cache, create new claimInfo object
-			// to be processed later.
-			claimInfo = newClaimInfoFromResourceClaim(resourceClaim)
-		}
-
-		// We delay checkpointing of this change until this call
-		// returns successfully. It is OK to do this because we
-		// will only return successfully from this call if the
-		// checkpoint has succeeded. That means if the kubelet is
-		// ever restarted before this checkpoint succeeds, the pod
-		// whose resources are being prepared would never have
-		// started, so it's OK (actually correct) to not include it
-		// in the cache.
-		claimInfo.addPodReference(pod.UID)
-
-		if claimInfo.prepared {
-			// Already prepared this claim, no need to prepare it again
-			continue
-		}
-
-		// Loop through all plugins and prepare for calling NodePrepareResources.
-		for _, resourceHandle := range claimInfo.ResourceHandles {
-			// If no DriverName is provided in the resourceHandle, we
-			// use the DriverName from the status
-			pluginName := resourceHandle.DriverName
-			if pluginName == "" {
-				pluginName = resourceClaim.Status.DriverName
+		// Add a defer to make sure we remove references to this pod in the
+		// claimInfo cache in cases where this function returns an error.
+		defer func(claim *resourceapi.ResourceClaim) {
+			if rerr != nil {
+				m.cache.Lock()
```

Comment:
Consider this sequence of events:
- PrepareResources is called
- locks the cache
- adds the pod
- somethings fails
- unlocks the cache
- *some other code runs, for example taking a snapshot*
- this `defer` runs, locks again and removes the pod

The other code in the middle sees the pod as being prepared. Isn't that wrong?

Reply:
I'm not not concerned with the sequence you mention, but there may still be an issue (see below):

Note that the `podUIDs` list does not track the preparation of the claim. It simply tracks which pods the kubelet is aware of as needing a claim to be prepared. And it only does so in order to ensure that all of the pods that reference the claim call unprepare before the kubelet finally unprepares the claim.

The the race I'd be worried about is if the unprepare call were to _not_ unprepare a resource that it was supposed to. 

So this sequence of events:

* PrepareResources is called by pod0
* locks the cache
* adds reference to pod0
* unlocks the cache
*  starts to prepare the resource

* PrepareResources is called by pod1
* locks the cache
* adds reference to pod1 
* unlocks the cache
*  starts to prepare the resource

* Pod 0 successfully prepares the claim
* locks the cache
* marks claim as prepared
* unlocks the cache

* UnprepareResources is called by pod0
* locks the cache
* sees two pod references
* removes the reference to pod 0
* unlocks the cache
* returns

* Pod 1 fails in the PrepareResources call
* the defer is called
* locks the cache
* removes the reference to pod 1
* unlocks the cache

* pod1 gets deleted before PrepareResources is called again
* UnprepareResources is *not* called for pod1 because it never had a successful PrepareResources call

Reply:
@klueska Would it help to only add pod reference when the claim is prepared?

Reply:
Here's another failure scenario:
- pod uses claims from two drivers
- PrepareResources creates claiminfo with the pod
- NodePrepareResources for driver A succeeds
- NodePrepareResources for driver B fails
- PrepareResources returns and removes the pod from the claiminfo, which is also not marked as "prepared"

If kubelet never retries PrepareResources, how does it know that NodeUnprepareResources must be called for driver A and when does it do that?

I suspect that `ClaimInfo.prepared` is too coarse. Can we turn that into:
- "not prepared" = definitely nothing is prepared
- "partially prepared" = zero or more NodePrepareResources calls were issued. This gets set in the cache *before* the first NodePrepareResources calls and checkpointed.
- "prepared" = all NodePrepareResources calls have completed, set and checkpointed at the end of NodePrepareResources

If PrepareResources gets called again for a partially prepared claim, it needs to do all NodePrepareResources calls again. We rely on drivers to detect that nothing needs to be done.

Same for UnprepareResources.

Is it guaranteed that one or the other gets called? In other words, can it be that we have stale entries in the cache that never get touched again?

I'd like to see some comments in the code that explain not just what PrepareResources do (I can see that in the code!) but also how and when they are called.

Reply:
@pohly 
> If kubelet never retries PrepareResources, how does it know that NodeUnprepareResources must be called for driver A and when does it do that?

To my understanding the `Kubelet` will retry `PrepareResources` if it fails, which is the case here.

Reply:
Will it retry if the function fails as describe above, kubelet gets stopped, the pod gets deleted, and kubelet gets restarted?

Reply:
No, it will not.

Reply:
It will probably not survive a node and a cluster deletion as well :)
It looks like we're trying to fix everything by this PR. Let me remind you that its original purpose was to fix these 2 bugs https://github.com/kubernetes/kubernetes/issues/123474 https://github.com/kubernetes/kubernetes/issues/116411.  I still believe that every revision of this PR was good enough to fix and backport. This doesn't mean that we should stop on that. Trying to come up with a perfect solution puts our users at risk of seeing these data races and checkpoint-related issues in 1.30.

Reply:
> Will it retry if the function fails as describe above, kubelet gets stopped, the pod gets deleted, and kubelet gets restarted?

That can be done on a Kubelet start if we have enough info/data checkpointed. Again, I'm not sure this PR is a right place to do it.

Reply:
Fixing 1.30 is a secondary priority. Our main priority has to be to get this code up to beta quality in 1.31 so that (ideally) we don't need to do redesigns for 1.32.

Reply:
True, but should it all be done in one PR that suppose to fix a couple of bugs?

Reply:
The title of this PR is "fix cache integrity". That is the purpose as far as I am concerned, and it currently doesn't achieve that because we have identified cases where the integrity is not ensured.

Reply:
I'd argue that this PR does fix cache integrity and worth merging into master and backporting. Calling UnprepareResources when the Kubelet is stopped and the pod is deleted is another issue and worth fixing by another PR. I don't fully agree that fixing data races in 1.30 is a second priority, considerting that people will try to use 1.30 for DRA experiments and the fact that we have plenty of time for 1.31 fixes.

Reply:
One thing that we still should considering doing in my opinion is to add pod references only when claim is prepared, i.e. when `NodePrepareResources` succeeds for all plugins. That can potentially fix the race that @klueska described above.

Reply:
Moving the addition of a pod reference until after the prepare call just flips the problem on its head. It introduces the possibility of an Unprepare call going through too early.

> PrepareResources is called by pod0
> NodePrepareResources succeeds for all plugins
> locks the cache
> adds reference to pod1
> checkpoints the cache
> unlocks the cache
> returns

> PrepareResources is called by pod1
> NodePrepareResources succeeds for all plugins
> (continued below ...)

> UnrepareResources is called by pod0
> NodePrepareResources succeeds for all plugins
> locks the cache
> removes reference to pod0
> NodeUnprepareResources succeeds for all plugins
> checkpoints the cache
> unlocks the cache
> returns

> (continued from above ...)
> locks the cache
> adds reference to pod1
> checkpoints the cache
> unlocks the cache
> returns

Now pod 1 thinks the claim is prepared, but it has actually been unprepared.

That said, we may be able to get away with this suggestion but it will require a bit more logic than simply moving the addition of the podReference until after the plugins have completed successfully.

Reply:
@klueska Locking claim info before calling `Node[Un]PrepareResources` and keeping it locked until it's fully prepared could be an acceptable trade-off. It would allow to work with other claim infos as cache is not always locked, but wouldn't allow calling GRPC for the same claim. WDYT?